File: app/src/main/java/com/esoxjem/movieguide/listing/sorting/SortingDialogPresenterImpl.java
package com.esoxjem.movieguide.listing.sorting;

/**
 * @author arun
 */
class SortingDialogPresenterImpl implements SortingDialogPresenter
{
    private SortingDialogView view;
    private SortingDialogInteractor sortingDialogInteractor;

    SortingDialogPresenterImpl(SortingDialogInteractor interactor)
    {
        sortingDialogInteractor = interactor;
    }

    @Override
    public void setView(SortingDialogView view)
    {
        this.view = view;
    }

    @Override
    public void destroy()
    {
        view = null;
    }

    @Override
    public void setLastSavedOption()
    {
        if (isViewAttached())
        {
            int selectedOption = sortingDialogInteractor.getSelectedSortingOption();

            if (selectedOption == SortType.MOST_POPULAR.getValue())
            {
                view.setPopularChecked();
            } else if (selectedOption == SortType.HIGHEST_RATED.getValue())
            {
                view.setHighestRatedChecked();
            } else if (selectedOption == SortType.NEWEST.getValue())
            {
                view.setNewestChecked();
            } else
            {
                view.setFavoritesChecked();
            }
        }
    }

    private boolean isViewAttached()
    {
        return view != null;
    }

    @Override
    public void onPopularMoviesSelected()
    {
        if (isViewAttached())
        {
            sortingDialogInteractor.setSortingOption(SortType.MOST_POPULAR);
            view.dismissDialog();
        }
    }

    @Override
    public void onHighestRatedMoviesSelected()
    {
        if (isViewAttached())
        {
            sortingDialogInteractor.setSortingOption(SortType.HIGHEST_RATED);
            view.dismissDialog();
        }
    }

    @Override
    public void onNewestMoviesSelected()
    {
        if (isViewAttached())
        {
            sortingDialogInteractor.setSortingOption(SortType.NEWEST);
            view.dismissDialog();
        }
    }

    @Override
    public void onFavoritesSelected()
    {
        if (isViewAttached())
        {
            sortingDialogInteractor.setSortingOption(SortType.FAVORITES);
            view.dismissDialog();
        }
    }
}

File: app/src/main/java/com/esoxjem/movieguide/movies/sorting/SortingDialogPresenterImpl.kt
package com.esoxjem.movieguide.movies.sorting

/**
 * @author arunsasidharan
 */
internal class SortingDialogPresenterImpl(private val sortingDialogInteractor: SortingDialogInteractor) : SortingDialogPresenter {
    private var view: SortingDialogView? = null

    override fun setView(view: SortingDialogView) {
        this.view = view
    }

    override fun destroy() {
        view = null
    }

    override fun setLastSavedOption() {
        val selectedOption = sortingDialogInteractor.getSelectedSortingOption()

        when (selectedOption) {
            SortType.MOST_POPULAR.value -> view!!.setPopularChecked()
            SortType.HIGHEST_RATED.value -> view!!.setHighestRatedChecked()
            SortType.NEWEST.value -> view!!.setNewestChecked()
            else -> view?.setFavoritesChecked()
        }
    }

    override fun onPopularMoviesSelected() {
        sortingDialogInteractor.setSortingOption(SortType.MOST_POPULAR)
        view?.dismissDialog()
    }

    override fun onHighestRatedMoviesSelected() {
        sortingDialogInteractor.setSortingOption(SortType.HIGHEST_RATED)
        view?.dismissDialog()
    }

    override fun onNewestMoviesSelected() {
        sortingDialogInteractor.setSortingOption(SortType.NEWEST)
        view?.dismissDialog()
    }

    override fun onFavoritesSelected() {
        sortingDialogInteractor.setSortingOption(SortType.FAVORITES)
        view?.dismissDialog()
    }
}
